src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -176,6 +182,10 @@ yield RunAcc{..} = do
       writeSTRef currentPageRef $! paEmpty
       pure $ Just (p, cs)
 
+countEntry :: RunAcc s -> ST s ()
+countEntry RunAcc{..} =
+    modifySTRef' entryCount $ \(NumEntries n) -> NumEntries (n + 1)
```

Comment:
Since this allocation of new `Int` heap objects happens for every single entry, it's a good use case for a `PrimVar` from the `primitive` package. Use `modifyPrimVar` in place of `modifySTRef`.

And the type would be `PrimVar s Int` rather than `STRef s NumEntries`. Though you could derive `Prim NumEntries` it's probably easier just to wrap at the point where we return results in `unsafeFinalise`.